src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,168 @@
-        return null;
+        checkIndex(index, size);
+        Node<T> currentNode = getCurrentNode(index);
+        T result = currentNode.value;
+        currentNode.value = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        T result = get(index);
+        unlink(getCurrentNode(index));
+        return result;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int index = 0; index < size; index++) {
+            if (areEqual(currentNode.value, object)) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addLastNode(Node<T> addedNode) {
+        tail.next = addedNode;
+        addedNode.prev = tail;
+        tail = addedNode;
+    }
+
+    private void checkIndex(int index, int bound) {
+        if (index < 0 || index >= bound) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of bounds");
+        }
+    }
+
+    private void addFirstNode(Node<T> addedNode) {
+        head.prev = addedNode;
+        addedNode.next = head;
+        addedNode.prev = null;
+        head = addedNode;
+    }
+
+    private void addMiddleNode(int index, Node<T> addedNode) {
+        Node<T> currentNode = getCurrentNode(index);
+        addedNode.next = currentNode;
+        addedNode.prev = currentNode.prev;
+        currentNode.prev.next = addedNode;
+        currentNode.prev = addedNode;
+    }
+
+    private Node<T> getCurrentNode(int index) {
+        checkIndex(index, size);
+        Node<T> currentNode;
+        if (index < size / 2) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private T unlink(Node<T> node) {
+        Node<T> next = node.next;
+        Node<T> prev = node.prev;
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+        T value = node.value;
+        node.value = null;
+        size--;
+        return value;
+    }
+
+    private boolean areEqual(T value1, T value2) {
+        return value1 == value2 || value1 != null && value1.equals(value2);
+    }
```

Comment:
I don't think you need this method to be extracted - you check for equality in a single place